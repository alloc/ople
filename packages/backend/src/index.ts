import './builtin'

export * from './serve'
export * from './fauna'
export * from './createQueries'
export * from './types'
