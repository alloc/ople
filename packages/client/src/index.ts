export { auto } from './auto'
export { watch } from './watch'
export { Ople, setup } from './Ople'
export { toDoc, toRef, trackChanges } from './OpleDocument'
export { OpleDate, OpleTime } from './values'
export { OplePages } from './OplePage'

// Advanced use
export * from './OpleBackend'
export * from './OpleContext'
export { onceCreated } from './OpleDocument'
export { makeCreator } from './creator'
export { makeSignal } from './signals'
export { attachAuto } from './auto'
export { OpleProtocol } from '@ople/agent'

// Observables
export * from 'wana'

// Types
export type { OpleRef } from './OpleRef'
export type { OpleDocument, OpleRefLike } from './OpleDocument'
export type { OpleSignal, OpleListener } from './signals'
export type { OplePage, OplePager } from './OplePage'
