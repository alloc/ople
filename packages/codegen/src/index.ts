export * from './server'
export * from './client'
export * from './backend'
export * from './parser'

export { warningsByFile } from './warnings'
