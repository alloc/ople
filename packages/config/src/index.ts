export { findConfig, loadConfig } from './loader'
export * from './projects'
export * from './fauna'
