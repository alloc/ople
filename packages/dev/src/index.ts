export * from './events'
export * from './dev'
// export * from './pushpin/start'
// export * from './pushpin/stop'
// export * from './fauna/start'
// export * from './fauna/stop'
