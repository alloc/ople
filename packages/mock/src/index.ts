import type { AgentFactory } from '@ople/agent'

export function mockBackend() {
  const send = jest.fn()
  const invoke = jest.fn()
  jest.mock('@ople/agent', (): { makeAgent: AgentFactory } => ({
    makeAgent: ({ host, port }) => ({
      host,
      port,
      sendQueue: [],
      send,
      invoke,
    }),
  }))
}
