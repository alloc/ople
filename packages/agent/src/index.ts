import { AgentConfig, config, OnSignal } from './config'
import { socket } from './socket'
import { send } from './send'

/** The backend communicator */
export const agent = {
  /** Update the agent configuration. */
  set(newConfig: AgentConfig) {
    if (socket) throw Error('Already connected')
    Object.assign(config, newConfig)
  },
  /** Send a message to the backend. */
  send,
}

/** Receive signals from the backend. */
export function onSignal(handler: OnSignal) {
  config.onSignal = handler
}
