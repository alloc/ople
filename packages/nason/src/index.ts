import { Encoder, use } from 'nason/src/index'
import arrayType from 'nason/src/types/array'
import objectType from 'nason/src/types/object'
import stringType from 'nason/src/types/string'
import { Nason, Packer, HandlePacker, PagerPacker } from './types'

export const makeNason = <
  OpleRef extends object,
  OpleTime extends object,
  OpleDate extends object,
  OpleHandle extends object,
  OplePager extends object
>(
  types: [
    OpleRef: Packer<OpleRef, string>,
    OpleTime: Packer<OpleTime, string>,
    OpleDate: Packer<OpleDate, string>,
    OpleHandle: HandlePacker<OpleHandle, OpleRef, OpleTime>,
    OplePager: PagerPacker<OplePager, OpleTime>
  ]
): Nason =>
  use([
    [0, encodeString(types[0])],
    [1, encodeString(types[1])],
    [2, encodeString(types[2])],
    [3, encodeHandle(types[3])],
    [4, encodePager(types[4])],
  ])

const encodeString = <T>({
  test,
  pack,
  unpack,
}: Packer<T, string>): Encoder<T> => ({
  test,
  encode: value => stringType.encode(pack(value)),
  decode: bytes => unpack(stringType.decode(bytes)),
})

const encodeHandle = <T extends object>({
  test = () => false,
  pack,
  unpack,
}: HandlePacker<T>): Encoder<T> => ({
  test,
  encode(handle, encode) {
    if (!pack) {
      throw Error('Document packing not supported')
    }
    const packed = pack(handle)
    const packedType: Encoder<any> = Array.isArray(packed)
      ? arrayType
      : objectType
    this.encode = handle => packedType.encode(pack(handle), encode)
    return packedType.encode(packed, encode)
  },
  decode(bytes, decode) {
    if (!unpack) {
      throw Error('Document unpacking not supported')
    }
    const isArray = bytes[0] === 5
    const packedType: Encoder<any> = isArray ? arrayType : objectType
    this.decode = bytes => unpack(packedType.decode(bytes, decode))
    return this.decode(bytes, decode)
  },
})

const encodePager = <T extends object>({
  test = () => false,
  pack,
  unpack,
}: PagerPacker<T>): Encoder<T> => ({
  test,
  encode(pager, encode) {
    if (!pack) {
      throw Error('Pager packing not supported')
    }
    return arrayType.encode(pack(pager), encode)
  },
  decode(bytes, decode) {
    if (!unpack) {
      throw Error('Pager unpacking not supported')
    }
    return unpack(arrayType.decode(bytes, decode) as any)
  },
})

export * from './batch'
export * from './reply'
export * from './types'
