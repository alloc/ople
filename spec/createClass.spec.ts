import { createClass, auto, OpleCreateFn, Ople, flushSync } from '../src'

describe('createClass', () => {
  it('creates a named constructor', () => {
    type State = { val: number; setVal(val: number): void }
    type Events = { change(val: number): void }

    const name = 'Test'
    const Test = createClass(
      name,
      (arg = 0): OpleCreateFn<State, Events> => (self, set, emit) => {
        set({ val: arg, setVal: val => set({ val }) })
        auto(() => emit('change', self.val))
      }
    )

    expect(Test.name).toBe(name)
    expect(Test.toString()).toMatchSnapshot()

    const test = new Test()
    expect(test instanceof Test).toBeTruthy()
    expect(test instanceof Ople).toBeTruthy()

    const str = test.toString()
    expect(str).toBe(new (class Test {})().toString())
    expect(str).toMatchInlineSnapshot('"[object Object]"')

    const onChange = jest.fn()
    test.on('change', onChange)
    test.setVal(1)
    flushSync()

    expect(onChange).toBeCalled()
    expect(onChange.mock.calls).toMatchSnapshot()
  })
})
