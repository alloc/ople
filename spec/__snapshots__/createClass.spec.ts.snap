// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createClass creates a named constructor 1`] = `
"function ctr() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var self = extendOple(Ople_1.Ople.call(this), create.apply(void 0, args));
        Object.setPrototypeOf(self, ctr.prototype);
        return self;
    }"
`;

exports[`createClass creates a named constructor 3`] = `
Array [
  Array [
    1,
  ],
]
`;
